PYME/recipes/kinetics_recipes.py

```diff
@@ -0,0 +1,428 @@
+        subtract one frame from each, giving us our final off times:
+
+        off times = [3, 1, 5, 3, 19]
+
+    **USE NOTE:
+        This module relies on good clustering. If clusters are incorrectly lumped together with the same index, this
+        will still go through with the subtraction, finding some off times between blinks from different molecules.
+        So, some off times will be incorrect and positive, which are impossible to filter out. Some off times will
+        be negative, which we can filter out.
+    """
+
+    inputName = Input('filtered_localizations')
+    outputName = Output('kinetics')
+
+    onTimesColName = CStr('on_times')
+    offTimesColName = CStr('off_times')
+
+    keepEndBlinks = Bool(True)
+
+
+    labelKey = CStr('dbscanClumpID')
+
+    def execute(self, namespace):
+
+        ds = tabular.mappingFilter(namespace[self.inputName])
+
+
+        last_frame = max(ds['t2'])+1
+
+        start_times = ds['t']
+        stop_times = ds['t2']
+
+        on_times = np.zeros_like(ds['t'])
+        off_times = np.zeros_like(ds['t'])
+
+        for i in np.unique(ds[self.labelKey]):
+
+            clump_data_idxs = np.where(ds[self.labelKey] == i)[0]
+
+            on_times[clump_data_idxs] = stop_times[clump_data_idxs] - start_times[clump_data_idxs] + 1
+
+            shifted_start_times = np.roll(start_times[clump_data_idxs], -1)
+
+            #TODO: this approach for the final off time might be wrong. OR would have to treat it differently as bleaached
+            cluster_stop_times = stop_times[clump_data_idxs]
+
+            if self.keepEndBlinks == True:
+                shifted_start_times[-1] = last_frame
+            elif self.keepEndBlinks == False:
+                shifted_start_times[-1] = cluster_stop_times[-1] + 1
+
+            loop_offtimes = shifted_start_times - cluster_stop_times - 1
+
+            off_times[clump_data_idxs] = np.round(loop_offtimes).astype('float')
+
+        try:
+            mdh = namespace[self.inputName].mdh        
+            ds.mdh = mdh
+            ds.mdh.setEntry('bad_offtimes.number_blinks_lost', len(np.where(off_times <= 0)[0]))
+            ds.mdh.setEntry('bad_offtimes.blinks_remaining', len(np.where(off_times > 0)[0]))
+        except:
+            pass
+
+        #todo: filter out bad points here, no point in blowing out non-optional filtering into unique modules, the whole thing will be cleaner then
+
+
+        ds.addColumn(self.onTimesColName, on_times)
+        ds.addColumn(self.offTimesColName, off_times)
+
+        namespace[self.outputName] = ds
+
+
+@register_module('FitSwitchingRates')
+class FitSwitchingRates(ModuleBase):
+    """
+    Calculates switching rates based on data which has been run through a blink finding script (trackutils.findtracks2),
+    a clustering script (hdbscan), and a script which calculates all unique on and off states (findblinkstatedurations)
```

Comment:
These docs could stand to be expanded quite a bit -> what's the model they're being fit to, etc. Ideally, you would include a general description of the fit, etc. up top, then type and details for each parameter. Preference is [numpy style](https://www.datacamp.com/community/tutorials/docstrings-python), and for recipe modules its nice to do a 'parameter' for each of the traits attributes.